src/main/scala/uk/gov/nationalarchives/Lambda.scala

```diff
@@ -0,0 +1,77 @@
+import pureconfig.generic.auto._
+import pureconfig.module.catseffect.syntax._
+import ujson.{Null, Value}
+import uk.gov.nationalarchives.Lambda.{Config, _}
+import uk.gov.nationalarchives.MetadataService.{AssetMetadata, BagitManifest, FileMetadata, FolderMetadata}
+import upickle.default
+import upickle.default._
+
+import java.io.{InputStream, OutputStream}
+import java.util.UUID
+
+class Lambda extends RequestStreamHandler {
+  val metadataService: MetadataService = MetadataService()
+  val dynamo: DADynamoDBClient[IO] = DADynamoDBClient[IO]()
+  val randomUuidGenerator: () => UUID = () => UUID.randomUUID()
+
+  implicit val inputReader: Reader[Input] = macroR[Input]
+  implicit val folderMetadataRowDecoder: CsvRowDecoder[FolderMetadata, String] = deriveCsvRowDecoder
+  implicit val assetMetadataRowDecoder: CsvRowDecoder[AssetMetadata, String] = deriveCsvRowDecoder
+  implicit val fileMetadataRowDecoder: CsvRowDecoder[FileMetadata, String] = deriveCsvRowDecoder
+  implicit val bagManifestRowDecoder: RowDecoder[BagitManifest] = deriveRowDecoder
+
+  implicit def OptionReader[T: Reader]: Reader[Option[T]] = reader[Value].map[Option[T]] {
+    case Null    => None
+    case jsValue => Some(read[T](jsValue))
+  }
+
+  implicit def OptionWriter[T: Writer]: Writer[Option[T]] = writer[Value].comap {
+    case Some(value) => write(value)
+    case None        => Null
+  }
+
+  override def handleRequest(inputStream: InputStream, outputStream: OutputStream, context: Context): Unit = {
+    val inputString = inputStream.readAllBytes().map(_.toChar).mkString
+    val input = read[Input](inputString)
+    for {
+      config <- ConfigSource.default.loadF[IO, Config]()
+      discoveryService <- DiscoveryService(config.discoveryApiUrl, randomUuidGenerator)
+      departmentAndSeries <- discoveryService.getDepartmentAndSeriesRows(input)
+      folderMetadata <- metadataService.parseCsvWithHeaders[FolderMetadata](input, "folder-metadata.csv")
+      assetMetadata <- metadataService.parseCsvWithHeaders[AssetMetadata](input, "asset-metadata.csv")
+      fileMetadata <- metadataService.parseCsvWithHeaders[FileMetadata](input, "file-metadata.csv")
+      bagManifests <- metadataService.parseCsvWithoutHeaders[BagitManifest](input, "manifest-sha256.txt")
+      entries <- metadataService.metadataToDynamoTables(input.batchId, departmentAndSeries, folderMetadata ++ assetMetadata ++ fileMetadata, bagManifests)
+      _ <- dynamo.writeItems(config.dynamoTableName, entries)
+    } yield {
+      val folderMetadataIds: List[UUID] = folderMetadata.filter(fm => fm.title != fm.name).map(_.identifier)
+      val departmentSeriesIds: List[UUID] = departmentAndSeries.department.id :: departmentAndSeries.series.map(_.id).toList
+
+      val stateData = StateData(
+        input.batchId,
+        input.s3Bucket,
+        input.s3Prefix,
+        folderMetadataIds ++ departmentSeriesIds,
+        folderMetadata.filter(fm => fm.title == fm.name).map(_.identifier),
```

Comment:
Since you have the opposite of what you did on line 56, here, do you think, on line 56, it's worth using `partitionMap` instead of filter
```
    val (folderMetadataIdsWhereTitleAndNameSame, folderMetadataIdsWhereTitleAndNameNotSame) =
      folderMetadata.partitionMap {
        case fm if fm.title == fm.name => Left(fm.identifier)
        case fmTitleAndNameDoNotMatch => Right(fmTitleAndNameDoNotMatch.identifier)
      }

Reply:
I don't know. It feels clearer as is at the expense of being slightly slower and I'd like to avoid constructing `Either` directly unless Scanamo makes me do it.

Reply:
Ok. Do you think `collect` might be better instead of a `filter` and `map`, so

` val folderMetadataIdsWhereTitleAndNameNotSame: List[UUID] = folderMetadata.collect{ fm if fm.title != fm.name => fm.identifier }`

and

`folderMetadata.collect{ case fm if fm.title != fm.name => fm.identifier }`?

Reply:
Ok, makes sense, I've changed it.